license_manager/service/service.go

```diff
@@ -2,36 +2,295 @@ package service
+		inv := &invocation{
+			ID:          invocationID,
+			Owner:       req.GetOwner(),
+			BuildTag:    req.GetBuildTag(),
+			LastCheckin: timeNow(),
+		}
+		lic.Enqueue(inv)
+
+		switch s.currentState {
+		case stateStarting:
+		case stateRunning:
+			lic.Promote()
+		default:
+			return nil, status.Errorf(codes.Internal, "unhandled state: %v", s.currentState)
+		}
+	}
+
+	// Invocation ID should now be known and either queued (due to the above
+	// insert, or from a previous request) or allocated (promoted from the queue
+	// by above, or asynchronously by the janitor).
+
+	if inv := lic.GetAllocated(invocationID); inv != nil {
+		// Invocation is allocated
+		inv.LastCheckin = timeNow()
+		return &lmpb.AllocateResponse{
+			ResponseType: &lmpb.AllocateResponse_LicenseAllocated{
+				LicenseAllocated: &lmpb.LicenseAllocated{
+					InvocationId:           invocationID,
+					LicenseRefreshDeadline: timestamppb.New(timeNow().Add(s.allocationRefreshDuration)),
+				},
+			},
+		}, nil
+	}
+	if inv := lic.GetQueued(invocationID); inv != nil {
+		// Invocation is queued
+		inv.LastCheckin = timeNow()
+		return &lmpb.AllocateResponse{
+			ResponseType: &lmpb.AllocateResponse_Queued{
+				Queued: &lmpb.Queued{
+					InvocationId: invocationID,
+					NextPollTime: timestamppb.New(timeNow().Add(s.queueRefreshDuration)),
+				},
+			},
+		}, nil
+	}
+	// Invocation is not allocated or queued
+	switch s.currentState {
+	case stateStarting:
+		// This invocation was previously queued before the server restart; add it
+		// back to the queue.
+		inv := &invocation{
+			ID:          invocationID,
+			Owner:       req.GetOwner(),
+			BuildTag:    req.GetBuildTag(),
+			LastCheckin: timeNow(),
+		}
+		lic.Enqueue(inv)
+		return &lmpb.AllocateResponse{
+			ResponseType: &lmpb.AllocateResponse_Queued{
+				Queued: &lmpb.Queued{
+					InvocationId: invocationID,
+					NextPollTime: timestamppb.New(timeNow().Add(s.queueRefreshDuration)),
+				},
+			},
+		}, nil
+	case stateRunning:
+		// This invocation is unknown (possibly expired)
+		return nil, status.Errorf(codes.FailedPrecondition, "invocation_id not found: %q", invocationID)
+	default:
+		return nil, status.Errorf(codes.Internal, "state not handled: %v", s.currentState)
+	}
 }
 
+// Refresh serves as a keepalive to refresh an allocation while an invocation
+// is still using it. See the proto docstrings for more info.
 func (s *Service) Refresh(ctx context.Context, req *lmpb.RefreshRequest) (*lmpb.RefreshResponse, error) {
-	return nil, status.Errorf(codes.Unimplemented, "Refresh() is not yet implemented")
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if len(req.GetLicenses()) != 1 {
+		return nil, status.Errorf(codes.InvalidArgument, "licenses must have exactly one license spec")
+	}
+	licenseType := formatLicenseType(req.GetLicenses()[0])
+	lic, ok := s.licenses[licenseType]
+	if !ok {
+		return nil, status.Errorf(codes.NotFound, "unknown license type: %q", licenseType)
+	}
+	invID := req.GetInvocationId()
```

Comment:
apart from locking, first ~10ish lines are common between Allocate and Refresh? Would it make sense to have a small helper? not sure myself, but maybe some of the "request validation" should be the same?

Reply:
This will disappear if we consolidate Refresh and Allocate.